shared/gateway/api_middleware.go

```diff
@@ -356,10 +360,102 @@ func (m *ApiProxyMiddleware) handleGetBeaconStateSsz(endpoint string, writer htt
+		e := fmt.Errorf("could not read response body: %w", err)
+		writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+		return
+	}
+
+	// Deserialize the output of grpc-gateway into the error struct.
+	errorJson := &DefaultErrorJson{}
+	if err := json.Unmarshal(body, errorJson); err != nil {
+		e := fmt.Errorf("could not unmarshal error: %w", err)
+		writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+		return
+	}
+	if errorJson.Msg() != "" {
+		// Something went wrong, but the request completed, meaning we can write headers and the error message.
+		for h, vs := range grpcResp.Header {
+			for _, v := range vs {
+				writer.Header().Set(h, v)
+			}
+		}
+		// Set code to HTTP code because unmarshalled body contained gRPC code.
+		errorJson.SetCode(grpcResp.StatusCode)
+		writeError(writer, errorJson, grpcResp.Header)
+		return
+		// Don't do anything if the response is only a status code.
+	}
+
+	// Deserialize the output of grpc-gateway.
+	responseJson := &BeaconStateSszResponseJson{}
+	if err := json.Unmarshal(body, responseJson); err != nil {
+		e := fmt.Errorf("could not unmarshal response: %w", err)
+		writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+		return
+	}
+	// Serialize the SSZ part of the deserialized value.
+	b, err := base64.StdEncoding.DecodeString(responseJson.Data)
+	if err != nil {
+		e := fmt.Errorf("could not decode response body into base64: %w", err)
+		writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+		return
+	}
+
+	// Write the response (headers + body) and PROFIT!
+	for h, vs := range grpcResp.Header {
+		for _, v := range vs {
+			writer.Header().Set(h, v)
+		}
+	}
+	writer.Header().Set("Content-Length", strconv.Itoa(len(b)))
+	writer.Header().Set("Content-Type", "application/octet-stream")
+	writer.Header().Set("Content-Disposition", "attachment; filename=beacon_state_ssz")
+	writer.WriteHeader(grpcResp.StatusCode)
+	if _, err := io.Copy(writer, ioutil.NopCloser(bytes.NewReader(b))); err != nil {
+		e := fmt.Errorf("could not write response message: %w", err)
+		writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+		return
+	}
+
+	// Final cleanup.
+	if err := grpcResp.Body.Close(); err != nil {
+		e := fmt.Errorf("could not close response body: %w", err)
+		writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+		return
+	}
+}
+
 func beaconStateSszRequested(endpoint string, request *http.Request) bool {
 	return endpoint == "/eth/v1/debug/beacon/states/{state_id}" && request.Header["Accept"][0] == "application/octet-stream"
 }
 
+func beaconBlockSszRequested(endpoint string, request *http.Request) bool {
+	return endpoint == "/eth/v1/beacon/blocks/{block_id}" && request.Header["Accept"][0] == "application/octet-stream"
```

Comment:
I imagine that a request could have multiple "accept" methods. Maybe abstract this check to iterate over all instances of the "Accept" header for a match.